src/Microsoft.DotNet.CodeFormatting/Rules/MarkReadonlyFieldsRule.cs

```diff
@@ -0,0 +1,527 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.DotNet.CodeFormatting.Rules
+{
+    /// <summary>
+    /// Mark any fields that can provably be marked as readonly.
+    /// </summary>
+    [GlobalSemanticRuleOrder(GlobalSemanticRuleOrder.MarkReadonlyFieldsRule)]
+    internal sealed class MarkReadonlyFieldsRule : IGlobalSemanticFormattingRule
+    {
+        private readonly SemaphoreSlim _processUsagesLock = new SemaphoreSlim(1, 1);
+        private ConcurrentDictionary<IFieldSymbol, bool> _unwrittenWritableFields;
+
+        public bool SupportsLanguage(string languageName)
+        {
+            return languageName == LanguageNames.CSharp;
+        }
+
+        public async Task<Solution> ProcessAsync(
+            Document document,
+            SyntaxNode syntaxRoot,
+            CancellationToken cancellationToken)
+        {
+            if (_unwrittenWritableFields == null)
+            {
+                using (await SemaphoreLock.GetAsync(_processUsagesLock))
```

Comment:
I can add a small optimization if the ConcurrentBag has a size of zero; since that would mean the code was conforming.  Then we don't actually have to rewrite anything, so at least all of the third walk could be skipped.  I'm not sure how to avoid the first two walks to actually tell if the code is conforming, though. The second could be skipped if there were no writable fields anywhere in the input, but that seems highly unlikely.